### PR TITLE
Initial implementation of NSSF operator

### PR DESCRIPTION
**Pull request by ghislainbourgeois**
# Description

Initial implementation of NSSF operator.

# Checklist:

- [x] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that validate the behaviour of the software
- [x] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: tests/unit/test_charm.py
Comment (PietroPasotti) [inline tests/unit/test_charm.py:1]: Not sure how strongly I feel about it, but in our team so far we've been keeping scenario tests clearly split from harness 'unit' tests.
We have them all in their own /tests/scenario folder. Especially if you plan to also use harness, it might be a good idea to keep them separated.

File: tests/unit/test_charm.py
Comment (PietroPasotti) [inline tests/unit/test_charm.py:11]: why did you have to type:ignore this line? issues with types?

File: tests/unit/test_charm.py
Comment (PietroPasotti) [inline tests/unit/test_charm.py:24]: it seems like in all of the tests there is a container and it can connect. Might be worth it putting this line in setUp, and whenever you need to modify the container (e.g. add Mounts), you can do:
```python
container = self.container.replace(mounts=<your mounts>)
state_in = State(containers=[container ...
```

File: tests/unit/test_charm.py
Comment (PietroPasotti) [inline tests/unit/test_charm.py:54]: nice!

File: tests/unit/test_charm.py
Comment (PietroPasotti) [inline tests/unit/test_charm.py:105]:
```suggestion
        patch_check_output.return_value = b"1.1.1.1"
```

File: tests/unit/test_charm.py
Comment (PietroPasotti) [inline tests/unit/test_charm.py:110]: ha! nice, I forgot they added this multi-with thing

File: tests/unit/test_charm.py
Comment (PietroPasotti) [inline tests/unit/test_charm.py:136]: why do you need this model name?

**Review by PietroPasotti — APPROVED** [review]
Looks very good! Only a few 'DRY' suggestions, where you could use scenario's `.replace` API to declare your objects once and obtain modified copies as needed, but the rest looks great!

File: lib/charms/sdcore_nrf/v0/fiveg_nrf.py
Comment (sanchezfdezjavier) [inline lib/charms/sdcore_nrf/v0/fiveg_nrf.py:113]: Please, fetch the new version of the lib. I had to re-add the `pytest-interface-tester` it was used in line `142`.

File: src/templates/nssfcfg.conf.j2
Comment (sanchezfdezjavier) [inline src/templates/nssfcfg.conf.j2:28]: Do we want to keep these configs?

File: tests/unit/test_charm.py
Comment (sanchezfdezjavier) [inline tests/unit/test_charm.py:55]: Following @gruyaume's comment on my last PR regarding multiple asserts in one test. I'd like to clarify when we should avoid them. This test seems reasonable, however, it can be split into one test for the waiting status and another one for the event being deferred. Let me know what you think!

File: tests/unit/test_charm.py
Comment (sanchezfdezjavier) [inline tests/unit/test_charm.py:59]: Are we adding more than one relation?

File: tests/unit/test_charm.py
Comment (sanchezfdezjavier) [inline tests/unit/test_charm.py:67]: How about extracting this `nrf_relation` to avoid some code duplication?

File: tests/unit/test_charm.py
Comment (sanchezfdezjavier) [inline tests/unit/test_charm.py:153]: Are we testing the pebble layer is added or whether if it's the expected one or not?

File: src/charm.py
Comment (sanchezfdezjavier) [inline src/charm.py:116]: From the docstring it might seem this method is doing two steps when is just pushing to the container, isn't it?
```suggestion
        """Pushes the NSSF config file to the container.
```

File: src/charm.py
Comment (sanchezfdezjavier) [inline src/charm.py:208]: Why do we need this ignore? I've never seen it before.

File: tests/integration/test_integration.py
Comment (PietroPasotti) [reply tests/integration/test_integration.py:41]: does it make sense to continue with the other tests if this one fails?

File: tests/unit/test_charm.py
Comment (PietroPasotti) [reply tests/unit/test_charm.py:67]: on the `.replace` comment I added elsewhere: you can do that with the State as well.
in `setUp`:

```python
self.state = State(containers=[self.container], leader=True, relations=[self.nrf_relation])
```

downside to that is that if you need to make a change to the container, you also need to modify the state: in the test:
```python
state = self.state.replace(relations=[self.relation.replace(...))
```

which is ugly. _I think_ I added `State.with_relations(...)` and `with_containers` to catch these common patterns though.

File: src/templates/nssfcfg.conf.j2
Comment (saltiyazan) [reply src/templates/nssfcfg.conf.j2:28]: probably they should come from the webui/simapp.
https://github.com/gruyaume/omec-simapp-operator/blob/main/src/files/default_config.yaml#L81

File: src/charm.py
Comment (saltiyazan) [inline src/charm.py:75]: Following [this comment](https://github.com/canonical/sdcore-amf-operator/pull/1#discussion_r1197067742) on the `amf` pr I think we can follow the same for all sdcore charms that need to push a config file and in case it does exist on the pod compare its content the one we are pushing and push if the content is different.

File: tests/unit/test_charm.py
Comment (saltiyazan) [reply tests/unit/test_charm.py:55]: In this test the condition is the same for both asserts, the test can fail only if that condition is unmet, regardless of the number of the asserts. (The asserts either will be true together or false together)
I believe it's a bad practice to have multiple asserts if it introduces a test case that can fail for multiple reasons. 
We should to test one thing at a time (and that is true here)
However it might become easier to read if split into two :)
That's my opinion @sanchezfdezjavier, but I'm curios what the others think too :)

File: src/charm.py
Comment (saltiyazan) [reply src/charm.py:75]: I see in the tests you are testing that the new content of a config is pushed if the content is changed, I assume the intended behaviour is to push anytime, and if it was changed it will be pushed anyway, which means that we are pushing the config file even when there is no change. 
Am I right? is this the intention?

File: metadata.yaml
Comment (gruyaume) [inline metadata.yaml:31]: How is this done, I see no reference to metrics/prometheus in the NSSF.

File: src/charm.py
Comment (gruyaume) [inline src/charm.py:176]: Probably shouldn't be a property

File: src/charm.py
Comment (gruyaume) [inline src/charm.py:187]: Probably shouldn't be a property

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py:43]: Probably not, I took that from another PR but I guess this is incomplete. I will check to see what is missing.

Comment (gruyaume) [reply src/charm.py:43]: I think it's mostly a matter of not supporting this integration.

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py:176]: Why?

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py:187]: Why?

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py:208]: Exactly, this is just telling the code coverage to ignore this block, as we will never test it in unit tests.

**Comment by gruyaume** [comment]
While we're here I'd also recommend:
- Standardising around the `SD-Core` naming (instead of `SD-CORE` or `SDCORE`)
- Use the `charming-actions/upload-charm` workflow

File: tests/unit/test_charm.py
Comment (ghislainbourgeois) [reply tests/unit/test_charm.py:1]: In this new charm, I was explicitly trying out Scenario as a replacement for harness. I am not expecting any uses of harness.

File: src/charm.py
Comment (gruyaume) [reply src/charm.py:176]: Well is it really an attribute of the charm? It feels like something we get from outside of the scope of the charm. Just the fact that we can't "set" it makes me think that it shouldn't be a property.

I'm to blame on this but overall it seems like we've overused properties a little bit.

File: tests/unit/test_charm.py
Comment (ghislainbourgeois) [reply tests/unit/test_charm.py:11]: Yes, we use `mypy`, and it complains that the package does not have library stubs or a `py.typed` marker. I think on the project you can simply add a `py.typed` file for the package to work with `mypy`.

File: tests/unit/test_charm.py
Comment (ghislainbourgeois) [reply tests/unit/test_charm.py:55]: I agree that tests should only test one thing, but this is different from having only one assertion. In this test case, I am validating that we are waiting. We are doing that by notifying the user through the status and deferring the event. We will never split those 2 behaviors.

In later tests, I have one that tests that the configuration is pushed and one that the pebble layer is applied. They are split in 2 tests because they are different things, even if the setup is exactly the same. We could very well move the configuration file handling to a different flow in the future.

File: tests/unit/test_charm.py
Comment (ghislainbourgeois) [reply tests/unit/test_charm.py:84]: It is a k8s charm. From the thread on MM, I see that `unit-get` is deprecated, and we should probably switch to `bind_address`. The [documentation](https://juju.is/docs/sdk/styleguide#heading--patterns) for that part is a bit confusing. It seems to require a relation, but I am not sure if any relation will do (I have only seen peer relations in the documentation for that). It also seem to be dependent on the state of the charm.

I would like nothing more than removing this patch. I am open to suggestions.

File: tests/unit/test_charm.py
Comment (ghislainbourgeois) [reply tests/unit/test_charm.py:153]: We are testing that the correct layer is applied.

File: tests/unit/test_charm.py
Comment (ghislainbourgeois) [reply tests/unit/test_charm.py]: Thanks, I updated to `replace`!

File: tests/unit/test_charm.py
Comment (ghislainbourgeois) [reply tests/unit/test_charm.py]: Thanks, applied!

File: lib/charms/sdcore_nrf/v0/fiveg_nrf.py
Comment (ghislainbourgeois) [reply lib/charms/sdcore_nrf/v0/fiveg_nrf.py]: Fetched, thanks!

File: src/templates/nssfcfg.conf.j2
Comment (ghislainbourgeois) [reply src/templates/nssfcfg.conf.j2]: I removed those configs, thanks!

File: tests/unit/test_charm.py
Comment (ghislainbourgeois) [reply tests/unit/test_charm.py]: I edited the name of the test.

File: tests/integration/test_integration.py
Comment (ghislainbourgeois) [reply tests/integration/test_integration.py:41]: In this case, it is the last test, so it does not change anything, but we will add more tests in the future, and they will not make sense if this one fails.

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py]: Thanks, fixed

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py]: Fixed

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py:176]: I think it can be seen as an attribute of the charm. In this case, the property has not setter, so it is a read-only property, and those are used all the time in Python.

I agree the value we get is not something inside the class or the instance, but I do not think it should prevent the use of property.

I think it makes those the intent of those kind of small lookup methods a bit clearer on use:

```
pod_ip = self._pod_ip

# vs

pod_ip = self._pod_ip()  # to me, this feels like it could be anything, and returning a value
```

I have tried to find opinions on that online, and I have not found anything. In the end, it is a super small detail, and I prefer their use, but I would be open to change the team's guidelines on this.

File: lib/charms/observability_libs/v0/juju_topology.py
Comment (gruyaume) [inline lib/charms/observability_libs/v0/juju_topology.py:1]: We won't need this either.

File: requirements.txt
Comment (gruyaume) [inline requirements.txt:5]: We need this in the charm?

File: src/charm.py
Comment (gruyaume) [inline src/charm.py:43]:
```suggestion
```

File: src/charm.py
Comment (gruyaume) [inline src/charm.py:24]:
```suggestion
```

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: src/templates/nssfcfg.conf.j2
Comment (gruyaume) [inline src/templates/nssfcfg.conf.j2:8]: Those would be provided by simapp/webui so if possible I'd remove them from here. It makes it confusing reading at config files if we see values that aren't really used.

Ref [here](https://github.com/gruyaume/omec-simapp-operator/blob/main/src/files/default_config.yaml#L89)

File: requirements.txt
Comment (ghislainbourgeois) [reply requirements.txt:5]: Yes, that is used by libs based on pydantic to validate the relation data.

File: lib/charms/observability_libs/v0/juju_topology.py
Comment (ghislainbourgeois) [reply lib/charms/observability_libs/v0/juju_topology.py:1]: Fixed

File: src/templates/nssfcfg.conf.j2
Comment (ghislainbourgeois) [reply src/templates/nssfcfg.conf.j2]: I removed the `snssai` section.

File: src/charm.py
Comment (gruyaume) [inline src/charm.py:5]:
```suggestion
"""Charmed operator for the SD-Core NSSF service."""
```

**Review by gruyaume — APPROVED** [review]
Approved, assuming we include the change for the `upload-charm` workflow and use the `SD-Core` naming.

File: tests/unit/test_charm.py
Comment (ghislainbourgeois) [reply tests/unit/test_charm.py]: I moved the complete relation to `setUp`, and create an empty relation in the only test requiring it.

@PietroPasotti, I checked for `with_relations` and `with_containers`, and it is not in the project yet. I don't dislike having the `State` explicitly defined in each test, it makes the `arrange` portion of the test clear. But redefining the same container, relations can get tiresome.

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py]: Updated the workflows, and the `SD-Core` naming.

File: src/charm.py
Comment (saltiyazan) [inline src/charm.py:82]: Don't we need to restart the workload when the pebble layer is added?
And do we need to add the layer even when it doesn't change?

File: src/charm.py
Comment (saltiyazan) [reply src/charm.py:69]: The config is passed to the startup command, I think we need to restart the workload to get the changes applied.

File: .github/workflows/publish-charm.yaml
Comment (saltiyazan) [inline .github/workflows/publish-charm.yaml:29]:
```suggestion
        uses: canonical/charming-actions/upload-charm@2.3.0
        with:
          built-charm-path: "sdcore-ausf_ubuntu-22.04-amd64.charm"
          credentials: "${{ secrets.CHARMCRAFT_AUTH }}"
```

File: src/charm.py
Comment (sanchezfdezjavier) [inline src/charm.py:49]: How about changing this method's signature to something more generic? This is not just handling pebble ready.

File: src/templates/nssfcfg.conf.j2
Comment (Gmerold) [inline src/templates/nssfcfg.conf.j2:23]: Do we need all these loggers here? Shouldn't each NF configure it's own logger?

File: src/charm.py
Comment (Gmerold) [inline src/charm.py:80]: Somewhat related to @saltiyazan comment - we should also restart a service when config changes but the pebble layer doesn't. We came up with a working solution in the [UDR charm](https://github.com/canonical/sdcore-udr-operator/blob/43e104901edd4d4c7eb2ec14001b3dc4702e7770/src/charm.py#L89)

File: src/charm.py
Comment (Gmerold) [inline src/charm.py:85]: Should we be enforcing named agrs for this function as we did in the SMF charm?

**Comment by Gmerold** [comment]
Please enable blocking merge if there are open convos.

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py]: I added a restart logic similar to what is in the UDR charm.

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py]: I added a restart logic similar to what is in the UDR charm.

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py:96]: Fixed

File: src/templates/nssfcfg.conf.j2
Comment (ghislainbourgeois) [reply src/templates/nssfcfg.conf.j2]: I removed all the loggers.

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py]: Fixed, thanks!

File: .github/workflows/publish-charm.yaml
Comment (gruyaume) [inline .github/workflows/publish-charm.yaml:29]:
```suggestion
          built-charm-path: "sdcore-nssf_ubuntu-22.04-amd64.charm"
```

File: src/charm.py
Comment (gruyaume) [inline src/charm.py:25]: This path doesn't match with the storage mount in `metadata.yaml`

**Review by gruyaume — CHANGES_REQUESTED** [review]

**Review by gruyaume — APPROVED** [review]

File: src/charm.py
Comment (ghislainbourgeois) [reply src/charm.py]: Fixed, thanks

**Pull request merged**
